vultr/resource_vultr_iso_private.go

```diff
@@ -107,7 +109,71 @@ func resourceVultrIsoDelete(ctx context.Context, d *schema.ResourceData, meta in
 	log.Printf("[INFO] Deleting iso : %s", d.Id())
 
 	if err := client.ISO.Delete(ctx, d.Id()); err != nil {
-		return diag.Errorf("error destroying ISO %s : %v", d.Id(), err)
+		// decode the error
+		var attachedErr struct {
+			Error  string `json:"error"`
+			Status int    `json:"status"`
+		}
+
+		errR := strings.NewReader(err.Error())
+		jsonErr := json.NewDecoder(errR).Decode(&attachedErr)
+		if jsonErr != nil {
+			return diag.Errorf("error destroying ISO %s : %v: %v", d.Id(), err, jsonErr)
+		}
+
+		if !strings.Contains(attachedErr.Error, "is still attached to") {
+			return diag.Errorf("error destroying ISO %s : %v: %v", d.Id(), err)
+		}
+
+		parts := strings.Split(attachedErr.Error, " ")
+		ip := parts[len(parts)-1]
+		if parsedIP := net.ParseIP(ip); parsedIP == nil {
+			return diag.Errorf("error destroying ISO %s : failed to parse IP to which ISO is attached", d.Id())
+		}
+
+		m := govultr.ListOptions{
```

Comment:
I feel like this section could be simplified a bit. What I would do is condense the two for loops you have into a single loop.

Also for the recursive call that could be replaced with just a iso.Delete call. If the wait succeeds with than the delete could be called. If the delete call fails for whatever reason just return an error.

```go
		// default is 100 instances
		options := &govultr.ListOptions{}
		for {
			instances, meta, err := client.Instance.List(ctx, options)
			if err != nil {
				return diag.Errorf("error destroying ISO %s : failed to list instances for detaching ISO", d.Id(), err)
			}

			for _, instance := range instances {
				if instance.MainIP == ip {
					if err := client.Instance.DetachISO(ctx, instance.ID); err != nil {
						return diag.Errorf("error destroying ISO %s : failed to detach from instances %s : %s", d.Id(), instance.ID, err)
					}

					if _, err := waitForIsoDetached(ctx, instance.ID, "ready", []string{"isomounted"}, "status", meta); err != nil {
						return diag.Errorf("error while waiting for ISO %s to detach from instance %s: %s", d.Id(), instance.ID, err)
					}

					if err := client.ISO.Delete(ctx, d.Id()); err != nil {
						return diag.Errorf("Error deleting ISO %s : %v", d.Id(), err)
					}
				}

				if meta.Links.Next == "" {
					break
				}
				options.Cursor = meta.Links.Next
			}
		}
	}
```

Reply:
Ah, thanks! I was looking for a nice way to loop through once. Super clean! I had to modify the above a little bit, since the `break` still lands in the infinite loop.

Please see https://github.com/vultr/terraform-provider-vultr/pull/131/commits/14e4628e6c817f068290584cfb14a2e0fa01c07a

Reply:
Ah yeah! my conditional  for the `meta.links.Next` was in the wrong loop.